File: packages/ui-native/src/components/Radio/RadioA11y.test.ts
import { describe, expect, it } from 'vitest';
import {
  getRadioAccessibilityProps,
  resolveSize,
  useRadioState,
} from './interface';

describe('radioA11y', () => {
  it('exposes role=radio with state.selected=true when checked', () => {
    const a11y = getRadioAccessibilityProps(
      { 'aria-label': 'Option A' },
      { isDisabled: false, isReadOnly: false, isChecked: true },
    );
    expect(a11y.accessible).toBe(true);
    expect(a11y.accessibilityRole).toBe('radio');
    expect(a11y.accessibilityLabel).toBe('Option A');
    expect(a11y.accessibilityState.selected).toBe(true);
    expect(a11y.accessibilityState.disabled).toBe(false);
  });

  it('reports state.selected=false when unchecked', () => {
    const a11y = getRadioAccessibilityProps(
      { 'aria-label': 'Option' },
      { isDisabled: false, isReadOnly: false, isChecked: false },
    );
    expect(a11y.accessibilityState.selected).toBe(false);
  });

  it('falls back to the visible label when aria-label is omitted', () => {
    const a11y = getRadioAccessibilityProps(
      { label: 'Standard delivery' },
      { isDisabled: false, isReadOnly: false, isChecked: false },
    );
    expect(a11y.accessibilityLabel).toBe('Standard delivery');
  });

  it('forwards aria-labelledby as accessibilityLabelledBy', () => {
    const a11y = getRadioAccessibilityProps(
      { 'aria-label': 'Box', 'aria-labelledby': 'group-heading' },
      { isDisabled: false, isReadOnly: false, isChecked: false },
    );
    expect(a11y.accessibilityLabelledBy).toBe('group-heading');
  });

  it('falls back to aria-describedby when labelledby is missing', () => {
    const a11y = getRadioAccessibilityProps(
      { 'aria-label': 'Box', 'aria-describedby': 'group-desc' },
      { isDisabled: false, isReadOnly: false, isChecked: false },
    );
    expect(a11y.accessibilityLabelledBy).toBe('group-desc');
  });

  it('hides the subtree when aria-hidden=true', () => {
    const a11y = getRadioAccessibilityProps(
      { 'aria-label': 'Box', 'aria-hidden': true },
      { isDisabled: false, isReadOnly: false, isChecked: false },
    );
    expect(a11y.accessible).toBe(false);
    expect(a11y.accessibilityElementsHidden).toBe(true);
    expect(a11y.importantForAccessibility).toBe('no-hide-descendants');
  });

  it('marks disabled state', () => {
    const a11y = getRadioAccessibilityProps(
      { 'aria-label': 'Box' },
      { isDisabled: true, isReadOnly: false, isChecked: false },
    );
    expect(a11y.accessibilityState.disabled).toBe(true);
  });

  it('passes accessibilityHint through unchanged', () => {
    const a11y = getRadioAccessibilityProps(
      { 'aria-label': 'Box', accessibilityHint: 'Selects this option' },
      { isDisabled: false, isReadOnly: false, isChecked: false },
    );
    expect(a11y.accessibilityHint).toBe('Selects this option');
  });
});

describe('useRadioState', () => {
  it('defaults size=m, appearance=secondary, isChecked=false', () => {
    const state = useRadioState({ value: 'x' });
    expect(state.resolvedSize).toBe('m');
    expect(state.resolvedAppearance).toBe('secondary');
    expect(state.isDisabled).toBe(false);
    expect(state.isReadOnly).toBe(false);
    expect(state.isChecked).toBe(false);
  });

  it('resolves auto appearance to secondary (matches web)', () => {
    const state = useRadioState({ value: 'x', appearance: 'auto' });
    expect(state.resolvedAppearance).toBe('secondary');
  });

  it('keeps explicit appearance', () => {
    const state = useRadioState({ value: 'x', appearance: 'positive' });
    expect(state.resolvedAppearance).toBe('positive');
  });

  it('derives isChecked directly from props.checked', () => {
    expect(useRadioState({ value: 'a', checked: true }).isChecked).toBe(true);
    expect(useRadioState({ value: 'a', checked: false }).isChecked).toBe(false);
    expect(useRadioState({ value: 'a' }).isChecked).toBe(false);
  });

  it('emits data attributes mirroring web (.module.css selectors)', () => {
    const unchecked = useRadioState({ value: 'x', size: 'l', appearance: 'negative' });
    expect(unchecked.dataAttrs).toEqual({
      'data-size': 'l',
      'data-appearance': 'negative',
      'data-unchecked': '',
    });

    const checked = useRadioState({ value: 'a', size: 'm', checked: true });
    expect(checked.dataAttrs).toMatchObject({ 'data-checked': '' });

    const ro = useRadioState({ value: 'a', readOnly: true, checked: true });
    expect(ro.dataAttrs).toMatchObject({ 'data-readonly': '', 'data-checked': '' });
  });

  it('forwards aria props on disabled / read-only', () => {
    expect(getRadioAccessibilityProps({}, { isDisabled: true, isReadOnly: false, isSelected: false })['aria-disabled']).toBe(true);
    expect(getRadioAccessibilityProps({}, { isDisabled: false, isReadOnly: true, isSelected: false })['aria-readonly']).toBe(true);
  });
});

describe('resolveSize', () => {
  it.each([
    ['s', 's'],
    ['m', 'm'],
    ['l', 'l'],
  ] as const)('passes %s through unchanged', (input, expected) => {
    expect(resolveSize(input)).toBe(expected);
  });

  it("falls back to 'm' for unknown / cast inputs", () => {
    expect(resolveSize('xl' as unknown as 'l')).toBe('m');
  });
});

File: apps/qa-playground-flutter/test/components/radio/radio_a11y_test.dart
/// Radio accessibility QA tests — resolver units + real widget semantics.
///
/// Asserts the merged [SemanticsData] the platform AT actually receives (role
/// via hasCheckedState + isInMutuallyExclusiveGroup, label, hint, enabled),
/// never a bare findsOneWidget for an a11y claim.
library;

import 'package:flutter/material.dart';
import 'package:flutter/semantics.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:ui_flutter/widgets/one_ui_radio.dart';
import 'package:ui_flutter/widgets/one_ui_radio_a11y.dart';
import 'package:ui_flutter/widgets/one_ui_radio_group.dart';

import '../../support/components/radio_harness.dart';

void main() {
  group('[a11y] resolveRadioAccessibilityLabel — RN label fallback chain', () {
    test('[a11y] prefers semanticsLabel over ariaLabel and visible label', () {
      expect(
        resolveRadioAccessibilityLabel(
          semanticsLabel: 'Screen reader',
          ariaLabel: 'Aria',
          label: 'Visible',
        ),
        'Screen reader',
      );
    });

    test('[a11y] prefers ariaLabel over visible label', () {
      expect(
        resolveRadioAccessibilityLabel(ariaLabel: 'Aria only', label: 'Visible'),
        'Aria only',
      );
    });

    test('[a11y] falls back to string child', () {
      expect(resolveRadioAccessibilityLabel(child: 'From child'), 'From child');
    });

    test('[a11y] returns null when nothing supplied', () {
      expect(resolveRadioAccessibilityLabel(), isNull);
    });
  });

  group('[a11y] getRadioAccessibilityProps — RN getRadioAccessibilityProps', () {
    test('[a11y] checked radio exposes selected + tappable', () {
      final a11y = getRadioAccessibilityProps(
        ariaLabel: 'Option',
        isDisabled: false,
        isReadOnly: false,
        isChecked: true,
      );
      expect(a11y.label, 'Option');
      expect(a11y.selected, isTrue);
      expect(a11y.canTap, isTrue);
      expect(a11y.exposeControl, isTrue);
    });

    test('[a11y] disabled blocks tap', () {
      final a11y = getRadioAccessibilityProps(
        ariaLabel: 'x',
        isDisabled: true,
        isReadOnly: false,
        isChecked: false,
      );
      expect(a11y.canTap, isFalse);
    });

    test('[a11y] readOnly exposes control but blocks tap', () {
      final a11y = getRadioAccessibilityProps(
        label: 'Locked',
        isDisabled: false,
        isReadOnly: true,
        isChecked: true,
      );
      expect(a11y.exposeControl, isTrue);
      expect(a11y.canTap, isFalse);
    });

    test('[a11y] aria-hidden collapses exposure', () {
      final a11y = getRadioAccessibilityProps(
        ariaLabel: 'Hidden',
        ariaHidden: true,
        isDisabled: false,
        isReadOnly: false,
        isChecked: false,
      );
      expect(a11y.hidden, isTrue);
      expect(a11y.exposeControl, isFalse);
    });

    test('[a11y] missing label defaults to "Radio"', () {
      final a11y = getRadioAccessibilityProps(
        isDisabled: false,
        isReadOnly: false,
        isChecked: false,
      );
      expect(a11y.label, 'Radio');
    });

    test('[a11y] forwards aria-describedby + hint', () {
      final a11y = getRadioAccessibilityProps(
        ariaLabel: 'x',
        ariaDescribedby: 'help-1',
        accessibilityHint: 'Select to choose',
        isDisabled: false,
        isReadOnly: false,
        isChecked: false,
      );
      expect(a11y.describedBy, 'help-1');
      expect(a11y.hint, 'Select to choose');
    });
  });

  group('[a11y] resolveOneUiRadioState — defaults + group inheritance', () {
    test('[a11y] defaults size=m, appearance=secondary, enabled', () {
      final s = resolveOneUiRadioState(isChecked: false);
      expect(s.resolvedSize, 'm');
      expect(s.resolvedAppearance, 'secondary');
      expect(s.isDisabled, isFalse);
      expect(s.isReadOnly, isFalse);
    });

    test('[a11y] auto resolves to secondary; explicit role wins', () {
      expect(resolveOneUiRadioState(appearance: 'auto', isChecked: false).resolvedAppearance,
          'secondary');
      expect(resolveOneUiRadioState(appearance: 'primary', isChecked: false).resolvedAppearance,
          'primary');
    });

    test('[a11y] group disabled/readOnly OR with option flags', () {
      final s = resolveOneUiRadioState(isChecked: false, groupDisabled: true);
      expect(s.isDisabled, isTrue);
      final r = resolveOneUiRadioState(isChecked: false, groupReadOnly: true);
      expect(r.isReadOnly, isTrue);
    });
  });

  group('[a11y] Radio widget — role + name + state', () {
    testWidgetsAllPlatforms('[a11y] checked option exposes the radio role (mutually exclusive)',
        (tester) async {
      await pumpRadioQaHarness(
        tester,
        radioInGroup([OneUiRadio(value: 'a', label: 'Accept')], defaultValue: 'a'),
      );
      withSemanticsHandle(tester, () {
        final d = radioSemanticsData(tester, 'Accept', checked: true);
        expect(d.hasFlag(SemanticsFlag.hasCheckedState), isTrue);
        expect(d.hasFlag(SemanticsFlag.isChecked), isTrue);
        expect(d.hasFlag(SemanticsFlag.isInMutuallyExclusiveGroup), isTrue);
      });
    });

    testWidgetsAllPlatforms('[a11y] label becomes the accessible name', (tester) async {
      await pumpRadioQaHarness(
        tester,
        radioInGroup([OneUiRadio(value: 'a', label: 'Accept')]),
      );
      withSemanticsHandle(tester, () {
        expect(radioSemanticsFinder('Accept'), findsWidgets);
      });
    });

    testWidgetsAllPlatforms('[a11y] ariaLabel overrides the visible label', (tester) async {
      await pumpRadioQaHarness(
        tester,
        radioInGroup([OneUiRadio(value: 'a', label: 'Visible', ariaLabel: 'From aria')]),
      );
      withSemanticsHandle(tester, () {
        expect(radioSemanticsFinder('From aria'), findsWidgets);
        expect(radioSemanticsFinder('Visible'), findsNothing);
      });
    });

    testWidgetsAllPlatforms('[a11y] disabled marks semantics not enabled', (tester) async {
      await pumpRadioQaHarness(
        tester,
        radioInGroup([OneUiRadio(value: 'a', label: 'Off')], disabled: true),
      );
      expectRadioDisabled(tester, 'Off');
    });

    testWidgetsAllPlatforms('[a11y] readOnly stays enabled in semantics', (tester) async {
      await pumpRadioQaHarness(
        tester,
        radioInGroup([OneUiRadio(value: 'a', label: 'Locked')], readOnly: true),
      );
      expectRadioEnabled(tester, 'Locked');
    });

    testWidgetsAllPlatforms('[a11y] accessibilityHint surfaces in semantics', (tester) async {
      await pumpRadioQaHarness(
        tester,
        radioInGroup([
          OneUiRadio(value: 'a', label: 'Hinted', accessibilityHint: 'Selects the plan'),
        ]),
      );
      withSemanticsHandle(tester, () {
        final d = radioSemanticsData(tester, 'Hinted');
        expect(d.hint, 'Selects the plan');
      });
    });

    testWidgetsAllPlatforms('[a11y] aria-hidden excludes the control from the tree',
        (tester) async {
      await pumpRadioQaHarness(
        tester,
        radioInGroup([OneUiRadio(value: 'a', label: 'Hidden', ariaHidden: true)]),
      );
      withSemanticsHandle(tester, () {
        expect(radioSemanticsFinder('Hidden'), findsNothing);
      });
    });

    testWidgetsAllPlatforms('[a11y] checked semantics flip on selection', (tester) async {
      await pumpRadioQaHarness(
        tester,
        radioInGroup([
          OneUiRadio(value: 'a', label: 'A'),
          OneUiRadio(value: 'b', label: 'B'),
        ], defaultValue: 'a'),
      );
      expectRadioChecked(tester, 'B', checked: false);
      await tester.tap(find.text('B'));
      await tester.pumpAndSettle();
      expectRadioChecked(tester, 'B', checked: true);
    });
  });

  group('[a11y] RadioGroup — radiogroup container role', () {
    testWidgetsAllPlatforms('[a11y] group exposes the radiogroup role', (tester) async {
      await pumpRadioQaHarness(
        tester,
        OneUiRadioGroup(
          semanticsLabel: 'Choose a plan',
          children: [
            OneUiRadio(value: 'a', label: 'A'),
            OneUiRadio(value: 'b', label: 'B'),
          ],
        ),
      );
      withSemanticsHandle(tester, () {
        final grp = find.descendant(
          of: radioGroupFinder(),
          matching: find.byWidgetPredicate(
              (w) => w is Semantics && w.properties.role == SemanticsRole.radioGroup),
        );
        expect(grp, findsOneWidget);
      });
    });
  });
}
